File: 백준/Gold/5582. 공통 부분 문자열/공통 부분 문자열.js
const input = require("fs").readFileSync("/dev/stdin", "utf-8").trim().split("\n");

const s1 = input[0].trim().split("");
const s2 = input[1].trim().split("");
const n1 = s1.length;
const n2 = s2.length;

const dp = Array.from({ length: n2 + 1 }, () => Array(n1 + 1).fill(0));
let answer = 0;

// 부분 문자열은 연속 해야됨 -> 연속하지 않으면 초기화!
for (let i = 1; i < n2 + 1; i++) {
	for (let j = 1; j < n1 + 1; j++) {
		if (s2[i - 1] === s1[j - 1]) {
			dp[i][j] = dp[i - 1][j - 1] + 1;
			if (dp[i][j] > answer) answer = dp[i][j];
		} else {
			dp[i][j] = 0;
		}
	}
}

console.log(answer);

File: 백준/Gold/5582. 공통 부분 문자열/공통 부분 문자열.py
s1 = list(input().strip())
n1 = len(s1)
s2 = list(input().strip())
n2 = len(s2)

dp = [[0] * (n1 + 1) for _ in range(n2 + 1)]
answer = 0

for i in range(1, n2 + 1):
    for j in range(1, n1 + 1):
        if s2[i - 1] == s1[j - 1]:
            dp[i][j] = dp[i - 1][j - 1] + 1
            if dp[i][j] > answer:
                answer = dp[i][j]
        else:
            dp[i][j] = 0
            
print(answer)